File: github_code/topological-sort/java/72.java
package com.pranav.hackerearth;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Queue;
import java.util.Scanner;

//https://www.hackerearth.com/practice/algorithms/graphs/topological-sort/practice-problems/algorithm/lonelyisland-49054110/description/
public class LonelyIsland {
	public static void main(String[] args) {
		Scanner s = new Scanner(System.in);
		int numberOfIslands = s.nextInt();
		int numberOfBridges = s.nextInt();
		int startingIsland = s.nextInt();
		startingIsland--;
		int[] indegree = new int[numberOfIslands];
		int[] outdegree = new int[numberOfIslands];
		Map<Integer, LinkedList<Integer>> adjList = new HashMap<>();
		for (int i = 0; i < numberOfIslands; i++) {
			adjList.put(i, new LinkedList<>());
		}
		int start;
		int end;
		for (int i = 0; i < numberOfBridges; i++) {
			start = s.nextInt() - 1;
			end = s.nextInt() - 1;
			indegree[end]++;
			outdegree[start]++;
			adjList.get(start).add(end);
		}
		s.close();
		Queue<Integer> islandWithZeroIndegree = new LinkedList<>();
		for (int i = 0; i < numberOfIslands; i++) {
			if (indegree[i] == 0) {
				islandWithZeroIndegree.add(i);
			}
		}
		double[] probabilty = new double[numberOfIslands];
		probabilty[startingIsland] = 1;
		while (!islandWithZeroIndegree.isEmpty()) {
			int current = islandWithZeroIndegree.remove();
			for (Integer i : adjList.get(current)) {
				probabilty[i] += (probabilty[current] * ((double)1 / (double)outdegree[current]));
				indegree[i]--;
				if (indegree[i] == 0) {
					islandWithZeroIndegree.add(i);
				}
			}

		}
		double max = 0;
		for (int i = 0; i < numberOfIslands; i++) {
			if (outdegree[i] == 0 && (probabilty[i] > max)) {
				max = probabilty[i];
			}
		}
		for(int i=0;i<numberOfIslands;i++) {
			if(outdegree[i]==0  && Math.abs(probabilty[i]-max) < 1e-9) {
				System.out.print(i+1+" ");
			}
		}
	}
}

File: github_code/topological-sort/cpp/72.c
//application of time stamping in dfs
//questiom link : https://www.hackerearth.com/practice/algorithms/graphs/topological-sort/practice-problems/algorithm/oliver-and-the-game-3/description/
#include <bits/stdc++.h>
using namespace std;

int c=0;
void dfs(vector <int >v[],int sn,bool vi[],int s[],int e[])
{
	vi[sn]=1;
	++c;
	s[sn]=c;
	for(int i=0;i<v[sn].size();i++)
	  {
	  	 if(vi[v[sn][i]]==0)
	  	   {
	  	   	  dfs(v,v[sn][i],vi,s,e);
	  	   	  
	  	   }
	  }
	++c;
	e[sn]=c;  
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
	long int n,i,x,y;
	cin>>n;
	vector <int >v[n+1];
	for(i=0;i<n-1;i++)
	  {
	  	 cin>>x>>y;
	  	 v[x].push_back(y);
	  	 v[y].push_back(x);
	  }
	 bool vi[n+1];
	 int s[n+1],e[n+1];
	 memset(vi,0,sizeof(vi));
	 memset(s,0,sizeof(s));
	 memset(e,0,sizeof(e));
	 dfs(v,1,vi,s,e);
	 int q;
	 /*for(i=1;i<n+1;i++)
	  {
	  	cout<<s[i]<<" "<<e[i]<<'\n';
	  }*/
	 
	 //cout<<s[8]<<" "<<e[8]<<'\n';
	 cin>>q;
	 //q=0;
             while(q--)
               {
                 cin>>i>>x>>y;
                 if(i==1)
                   {
                   	swap(x,y);
                   }
                  if(s[x]<=s[y] && e[y]<=e[x])
                   	  {
                   	  	cout<<"YES"<<'\n';
                   	  	
                   	  }
                   	  else
                   	    cout<<"NO"<<'\n'; 
               }
	return 0;
}
